### PR TITLE
[risk-low][RW-7060] Show a toast-banner when the user is newly eligible for Controlled Tier 

### PR DESCRIPTION
**Pull request by jmthibault79**
CT Available

Clicking X closes the toast-banner and stores a cookie, preventing the user from ever seeing this again.

Clicking `complete the necessary steps` closes the toast-banner and routes to the DAR page.

Only triggers when all of the following conditions are met.
- the env allows users to see the CT in the UI
- the user does not have CT membership
- the user is CT-eligible
- the user’s first sign in time is BEFORE the creation time of the latest CT CDR
- the user has not yet acked the notification previously

Tested locally.

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/main/resources/workbench-api.yaml [hunk 1]
```diff
@@ -4043,7 +4044,12 @@ definitions:
         type: string
       defaultCdrVersionId:
         type: string
+      defaultCdrVersionCreationTime:
+        type: integer
         format: int64
```
Comment (jmthibault79) [inline api/src/main/resources/workbench-api.yaml:4049]: this is funny because it actually is a string representation of an int64

File: ui/src/app/components/toast-banner.tsx [hunk 2]
```diff
@@ -73,15 +73,15 @@ interface ToastProps {
   title: string;
   message: string;
   onClose: Function;
-  type: ToastType;
```
Comment (jmthibault79) [inline ui/src/app/components/toast-banner.tsx:76]: naming something `type` is just asking for trouble.  Let's avoid that trouble.

File: ui/src/app/components/status-alert-banner-maybe.tsx [hunk 3]
```diff
@@ -41,7 +41,7 @@ export const StatusAlertBannerMaybe = () => {
     getAlert();
   }, []);
 
-  const onStatusAlertBannerUnmount = () => {
+  const acknowledgeAlert = () => {
```
Comment (jmthibault79) [inline ui/src/app/components/status-alert-banner-maybe.tsx:44]: not really related to other changes, just thought it was a better name

File: ui/src/app/components/ct-available-banner-maybe.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  it('should render if all of the requirements are met', () => {
+    fulfillAllBannerRequirements();
+
+    const wrapper = component();
+    expect(wrapper.exists()).toBeTruthy();
+
+    expect(ctBannerExists(wrapper)).toBeTruthy();
+  });
+
+  it('should not render if all of the requirements are met, except that the user is not CT eligible', () => {
```
Comment (jmthibault79) [inline ui/src/app/components/ct-available-banner-maybe.spec.tsx:105]: Please review this testing approach.  After approval, I'll add tests for the other negative cases.

File: ui/src/app/components/ct-available-banner-maybe.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  // all of the following must be true
+  const shouldShow =
+    // the user is eligible for the CT
+    eligibleForTier(profile, AccessTierShortNames.Controlled) &&
+    // the user does not currently have CT access
+    !profile.accessTierShortNames.includes(AccessTierShortNames.Controlled) &&
+    // the environment allows users to see the CT (in the UI)
+    environment.accessTiersVisibleToUsers.includes(AccessTierShortNames.Controlled) &&
+    // the user's DUCC access module completion time was before the release of the default CT CDR Version
+    (getAccessModuleStatusByName(profile, AccessModule.DATAUSERCODEOFCONDUCT).completionEpochMillis <
```
Comment (jmthibault79) [inline ui/src/app/components/ct-available-banner-maybe.tsx:26]: TODO: what about bypass?

File: ui/src/app/components/toast-banner.tsx [hunk 6]
```diff
@@ -71,17 +71,17 @@ const styleForType = (tt: ToastType) => switchCase(tt,
 
 interface ToastProps {
   title: string;
-  message: string;
+  message: string | JSX.Element;
```
Comment (jmthibault79) [inline ui/src/app/components/toast-banner.tsx:74]: TODO test the expanded types in other contexts

File: ui/src/app/pages/login/account-creation/account-creation.spec.tsx [hunk 7]
```diff
@@ -65,7 +65,7 @@ it('should handle username validity ends with .', () => {
   expect(wrapper.exists('#usernameError')).toBeTruthy();
 });
 
-test.each(['user@name', '.user', 'user.', 'user..', "O'Riley", '50%', 'no+plus', 'ælfred', 'money$man'])
+test.each(['user@name', '.user', 'user.', 'user..', 'O\'Riley', '50%', 'no+plus', 'ælfred', 'money$man'])
```
Comment (jmthibault79) [inline ui/src/app/pages/login/account-creation/account-creation.spec.tsx:68]: the linter always wins eventually

File: ui/src/app/components/toast-banner.tsx (hunk 6 above)
Comment (jmthibault79) [reply ui/src/app/components/toast-banner.tsx:74]: tested locally 👍

File: api/src/test/java/org/pmiops/workbench/cdr/CdrVersionServiceTest.java [hunk 8]
```diff
@@ -321,42 +319,40 @@ private void testGetCdrVersionsHasDataType(Predicate<CdrVersion> hasType) {
...
+  private List<CdrVersion> parseTierVersions(
+      CdrVersionTiersResponse cdrVersionsByTier, String accessTierShortName) {
+    return parseTier(cdrVersionsByTier, accessTierShortName).getVersions();
   }
 
   private DbCdrVersion makeCdrVersion(
       long cdrVersionId,
       boolean isDefault,
       String name,
-      long creationTime,
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/cdr/CdrVersionServiceTest.java:337]: some of these are not needed for this test

File: api/src/test/java/org/pmiops/workbench/cdr/CdrVersionServiceTest.java [hunk 9]
```diff
@@ -283,35 +272,44 @@ public void testGetCdrVersionsHasWgsData() {
...
   private void testGetCdrVersionsHasDataType(Predicate<CdrVersion> hasType) {
     addMembershipForTest(registeredTier);
     final List<CdrVersion> cdrVersions =
-        parseRegisteredTier(cdrVersionService.getCdrVersionsByTier());
+        parseTierVersions(cdrVersionService.getCdrVersionsByTier(), registeredTier.getShortName());
     // hasFitBitData, hasCopeSurveyData, hasMicroarrayData, and hasWgsData are false by default
     assertThat(cdrVersions.stream().anyMatch(hasType)).isFalse();
 
-    makeCdrVersion(
-        3L, true, "Test CDR With Data Types", 123L, registeredTier, "wgs", true, true, "", "");
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/cdr/CdrVersionServiceTest.java:312]: 3L is no longer an available ID at this point

File: api/src/main/java/org/pmiops/workbench/cdr/CdrVersionService.java [hunk 10]
```diff
@@ -99,11 +104,24 @@ private CdrVersionTier getVersionsForTier(DbAccessTier accessTier) {
...
+    CdrVersion defaultVersion =
+        cdrVersionMapper.dbModelToClient(getDefaultVersionForTier(accessTier, cdrVersions));
+
+    return new CdrVersionTier()
+        .versions(
+            cdrVersions.stream()
+                .map(cdrVersionMapper::dbModelToClient)
+                .collect(Collectors.toList()))
+        .defaultCdrVersionId(defaultVersion.getCdrVersionId())
+        .defaultCdrVersionCreationTime(defaultVersion.getCreationTime())
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/cdr/CdrVersionService.java:116]: newly added for this feature - it affects whether we show the banner

File: ui/src/app/routing/guards.tsx [hunk 11]
```diff
@@ -13,7 +13,7 @@ export const signInGuard: Guard = {
 };
 
 const userIsEnabled = (userDisabledInDb: boolean) =>
-    (!userDisabledInDb && eligibleForRegisteredTier(profileStore.get().profile.tierEligibilities));
+  (!userDisabledInDb && eligibleForTier(profileStore.get().profile, AccessTierShortNames.Registered));
```
Comment (NehaBroad) [inline ui/src/app/routing/guards.tsx:16]: I just found out about  hasRegisteredTierAccess in access-tiers.tsx. Maybe we can have a ticket later to use just one

File: ui/src/app/components/ct-available-banner-maybe.tsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  const ct = cdrVersionTiers?.find(v => v.accessTierShortName === AccessTierShortNames.Controlled);
+
+  // all of the following must be true
+  const shouldShow = profile && ct &&
+    // the environment allows users to see the CT (in the UI)
+    environment.accessTiersVisibleToUsers.includes(AccessTierShortNames.Controlled) &&
+    // the user is eligible for the CT
+    eligibleForTier(profile, AccessTierShortNames.Controlled) &&
+    // the user does not currently have CT access
+    !profile.accessTierShortNames.includes(AccessTierShortNames.Controlled) &&
```
Comment (NehaBroad) [inline ui/src/app/components/ct-available-banner-maybe.tsx:27]: How is  eligibleForTier(profile, AccessTierShortNames.Controlled) and !profile.accessTierShortNames.includes(AccessTierShortNames.Controlled) different?

Comment (NehaBroad) [reply ui/src/app/components/ct-available-banner-maybe.tsx:27]: nvm

File: ui/src/app/components/ct-available-banner-maybe.tsx [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+  const acknowledgeBanner = () => {
+    if (cookiesEnabled()) {
+      localStorage.setItem(CT_COOKIE_KEY, 'acknowledged');
+    }
+    setShowBanner(false);
+  };
+
+  const bannerBody = <div data-test-id='controlled-tier-available'>
+    To access genomic data and more in the <i>All of Us</i> controlled tier,
```
Comment (NehaBroad) [inline ui/src/app/components/ct-available-banner-maybe.tsx:56]: There is <AoU> or something that does <i>All of Us</i>

File: ui/src/app/components/ct-available-banner-maybe.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+  }
+  const userIsNew = () => {
+    updateProfile({newFirstSignIn: TIME3});
+  }
+  const darActive = () => {
+    window.location.pathname = DAR_PATH;
+  }
+
+  test.each([
+    ['there is no visible CT', ctNotVisible, () => {}],
```
Comment (NehaBroad) [inline ui/src/app/components/ct-available-banner-maybe.spec.tsx:124]: nice!

**Review by NehaBroad — APPROVED** [review]

File: ui/src/app/routing/guards.tsx (hunk 11 above)
Comment (jmthibault79) [reply ui/src/app/routing/guards.tsx:16]: This is slightly different.  It checks whether a user **can** have access vs. whether a user **does** have access.

File: ui/src/app/components/ct-available-banner-maybe.tsx (hunk 13 above)
Comment (jmthibault79) [reply ui/src/app/components/ct-available-banner-maybe.tsx:56]: right, `<AoU/>`.  I'll update.
